src/runtime/library_module.cc

```diff
@@ -123,20 +126,63 @@ void ImportModuleBlob(const char* mblob, std::vector<Module>* mlist) {
   dmlc::Stream* stream = &fs;
   uint64_t size;
   CHECK(stream->Read(&size));
+  std::vector<Module> modules;
+  bool has_import_tree = false;
   for (uint64_t i = 0; i < size; ++i) {
     std::string tkey;
     CHECK(stream->Read(&tkey));
-    if (tkey == "c") continue;
-    std::string fkey = "module.loadbinary_" + tkey;
-    const PackedFunc* f = Registry::Get(fkey);
-    CHECK(f != nullptr)
+    // Currently, _lib is for DSOModule, but we
+    // don't have loadbinary function for it currently
+    if (tkey == "_lib") {
+      auto dso_module = Module(make_object<LibraryModuleNode>(lib));
+      // allow lookup of symbol from dso root (so all symbols are visible).
+      if (auto *ctx_addr =
+          reinterpret_cast<void**>(lib->GetSymbol(runtime::symbol::tvm_module_ctx))) {
+        *ctx_addr = dso_module.operator->();
+      }
+      modules.emplace_back(dso_module);
+    } else if (tkey == "_import_tree") {
+      has_import_tree = true;
+      std::vector<uint64_t> import_tree_row_ptr;
+      std::vector<uint64_t> import_tree_child_indices;
+      CHECK(stream->Read(&import_tree_row_ptr));
+      CHECK(stream->Read(&import_tree_child_indices));
+      for (size_t i = 0; i < modules.size(); i++) {
+        for (size_t j = import_tree_row_ptr[i]; j < import_tree_row_ptr[i + 1]; j++) {
+          auto module_import_addr = ModuleInternal::GetImportsAddr(modules[i].operator->());
+          module_import_addr->emplace_back(modules[import_tree_child_indices[j]]);
+        }
+      }
+    } else {
+      std::string fkey = "module.loadbinary_" + tkey;
+      const PackedFunc* f = Registry::Get(fkey);
+      CHECK(f != nullptr)
         << "Loader of " << tkey << "("
         << fkey << ") is not presented.";
-    Module m = (*f)(static_cast<void*>(stream));
-    mlist->push_back(m);
+      Module m = (*f)(static_cast<void*>(stream));
+      modules.emplace_back(m);
+    }
+  }
+  // if we are using old dll, we don't have import tree
+  // so that we can't reconstruct module relationship using import tree
+  if (!has_import_tree) {
+    auto n = make_object<LibraryModuleNode>(lib);
+    auto module_import_addr = ModuleInternal::GetImportsAddr(n.operator->());
+    for (const auto& m : modules) {
+      module_import_addr->emplace_back(m);
+    }
+    // allow lookup of symbol from dso root (so all symbols are visible).
+    if (auto *ctx_addr =
+        reinterpret_cast<void**>(lib->GetSymbol(runtime::symbol::tvm_module_ctx))) {
+      *ctx_addr = n.operator->();
+    }
+    return Module(n);
   }
+  CHECK(!modules.empty());
+  return modules[0];
```

Comment:
Add comment: invariance: root module is always at location 0. The module order is collected via DFS. Add the requirement comment to the serializer as well.